File: agent-guard-plugin/lib/gates.js
const fs = require("node:fs");
const path = require("node:path");
const { loadState, AGENT_DIR } = require("./state");
const { loadJobs } = require("./jobs");
const { loadPlanSummary } = require("./plan");

function canFinalize(rootDir) {
  const state = loadState(rootDir);
  const jobs = loadJobs(rootDir);
  const plan = loadPlanSummary(rootDir);
  const reasons = [];

  if (state.remaining_steps.length > 0) {
    reasons.push("remaining_steps is not empty");
  }

  if (jobs.jobs.some((job) => job.status === "running")) {
    reasons.push("running jobs still exist");
  }

  if (!state.last_verification || state.last_verification.exit_code !== 0) {
    reasons.push("latest final verification is missing or failed");
  }

  if (plan.includesReview) {
    const reviewPath = path.join(rootDir, AGENT_DIR, "artifacts", "review.json");
    if (!fs.existsSync(reviewPath)) {
      reasons.push("review artifact is required by plan but missing");
    }
  }

  if (state.can_finalize !== true) {
    reasons.push("state.can_finalize is not true");
  }

  if (reasons.length > 0) {
    return {
      decision: "block",
      reasons,
    };
  }

  return {
    decision: "allow",
    reasons: [],
  };
}

module.exports = {
  canFinalize,
};

File: agent-guard-plugin/src/agent_guard/gates.py
from __future__ import annotations

from pathlib import Path

from .jobs import load_jobs
from .plan import load_plan, nonterminal_plan_steps
from .state import load_state


def can_finalize(root_dir: Path) -> dict[str, object]:
    # Finalization is intentionally lightweight: no running jobs, the workflow
    # has explicitly enabled finalization, and any plan steps are terminal.
    state = load_state(root_dir)
    jobs = load_jobs(root_dir)
    reasons: list[str] = []

    if any(job.get("status") == "running" for job in jobs.get("jobs", [])):
        reasons.append("running jobs still exist")

    if state.get("can_finalize") is not True:
        reasons.append("state.can_finalize is not true")

    if load_plan(root_dir) is not None:
        nonterminal_steps = nonterminal_plan_steps(root_dir)
        if nonterminal_steps:
            names = ", ".join(step["name"] for step in nonterminal_steps)
            reasons.append(f"plan.yaml has non-terminal steps: {names}")

    if reasons:
        return {"decision": "block", "reasons": reasons}

    return {"decision": "allow", "reasons": []}
